crates/re_log_types/src/data_table_batcher.rs

```diff
@@ -0,0 +1,787 @@
+    ///
+    /// This does **not** wait for the flush to propagate (see [`Self::flush_blocking`]).
+    /// See [`DataTableBatcher`] docs for ordering semantics and multithreading guarantees.
+    #[inline]
+    pub fn flush_async(&self) {
+        self.inner.flush_async();
+    }
+
+    /// Initiates a flush the batching pipeline and waits for it to propagate.
+    ///
+    /// See [`DataTableBatcher`] docs for ordering semantics and multithreading guarantees.
+    #[inline]
+    pub fn flush_blocking(&self) {
+        self.inner.flush_blocking();
+    }
+
+    // --- Subscribe to tables ---
+
+    /// Returns a _shared_ channel in which are sent the batched [`DataTable`]s.
+    ///
+    /// Shutting down the batcher will close this channel.
+    ///
+    /// See [`DataTableBatcher`] docs for ordering semantics and multithreading guarantees.
+    pub fn tables(&self) -> Receiver<DataTable> {
+        self.inner.rx_tables.clone()
+    }
+}
+
+impl DataTableBatcherInner {
+    fn push_row(&self, row: DataRow) {
+        self.send_cmd(Command::AppendRow(row));
+    }
+
+    fn flush_async(&self) {
+        let (flush_cmd, _) = Command::flush();
+        self.send_cmd(flush_cmd);
+    }
+
+    fn flush_blocking(&self) {
+        let (flush_cmd, oneshot) = Command::flush();
+        self.send_cmd(flush_cmd);
+        oneshot.recv().ok();
+    }
+
+    fn send_cmd(&self, cmd: Command) {
+        // NOTE: Internal channels can never be closed outside of the `Drop` impl, this cannot
+        // fail.
+        self.tx_cmds.send(cmd).ok();
+    }
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn batching_thread(
+    config: DataTableBatcherConfig,
+    rx_cmd: Receiver<Command>,
+    tx_table: Sender<DataTable>,
+) {
+    let rx_tick = crossbeam::channel::tick(config.flush_tick);
+
+    struct Accumulator {
+        latest: Instant,
+        pending_rows: Vec<DataRow>,
+        pending_num_rows: u64,
+        pending_num_bytes: u64,
+    }
+
+    impl Accumulator {
+        fn reset(&mut self) {
+            self.latest = Instant::now();
+            self.pending_rows.clear();
+            self.pending_num_rows = 0;
+            self.pending_num_bytes = 0;
+        }
+    }
+
+    let mut acc = Accumulator {
+        latest: Instant::now(),
+        pending_rows: Default::default(),
+        pending_num_rows: Default::default(),
+        pending_num_bytes: Default::default(),
+    };
+
+    fn do_push_row(
+        config: &DataTableBatcherConfig,
+        acc: &mut Accumulator,
+        mut row: DataRow,
+    ) -> bool {
+        // TODO(#1760): now that we're re doing this here, it really is a massive waste not to send
+        // it over the wire...
+        row.compute_all_size_bytes();
+
+        acc.pending_num_rows += 1;
+        acc.pending_num_bytes += row.total_size_bytes();
+        acc.pending_rows.push(row);
+
+        acc.pending_num_rows >= config.flush_num_rows
+            || acc.pending_num_bytes >= config.flush_num_bytes
```

Comment:
Rather than returning a bool it seems like it would be clearer to call `do_flush_all` here.

Reply:
I find it clearer to be able to see all flush triggers in the main loop

Reply:
In that case why not move the *check* into the main loop as well?
```
do_push_row(&mut acc, row);
if acc.pending_num_rows >= config.flush_num_rows || acc.pending_num_bytes >= config.flush_num_bytes
{
    do_flush_all(&mut acc, &tx_table, "bytes|rows");
    acc.reset();
}
```

But, at the very least add a comment to `do_push_row` indicating that it returns a bool with the assumption that the caller will flush the data if it returns true?

Reply:
Moving the check sounds good to me :+1: